base_tier_validation/models/tier_validation.py

```diff
@@ -0,0 +1,194 @@
+                    reviews = rec.request_validation()
+                    rec._validate_tier(reviews)
+                    if not self._calc_reviews_validated(reviews):
+                        raise ValidationError(_(
+                            "This action needs to be validated for at least "
+                            "one record. \nPlease request a validation."))
+                if rec.review_ids and not rec.validated:
+                    raise ValidationError(_(
+                        "A validation process is still open for at least "
+                        "one record."))
+            if (rec.review_ids and getattr(rec, self._state_field) in
+                    self._state_from and not vals.get(self._state_field) in
+                    (self._state_to + [self._cancel_state]) and not
+                    self._check_allow_write_under_validation(vals)):
+                raise ValidationError(_("The operation is under validation."))
+        if vals.get(self._state_field) in self._state_from:
+            self.mapped('review_ids').unlink()
+        return super(TierValidation, self).write(vals)
+
+    def _validate_tier(self, tiers=False):
+        self.ensure_one()
+        tier_reviews = tiers or self.review_ids
+        user_reviews = tier_reviews.filtered(
+            lambda r: r.status in ('pending', 'rejected') and
+            (r.reviewer_id == self.env.user or
+             r.reviewer_group_id in self.env.user.groups_id))
+        user_reviews.write({'status': 'approved'})
+
+    @api.multi
+    def validate_tier(self):
+        for rec in self:
+            rec._validate_tier()
+
+    @api.multi
+    def reject_tier(self):
+        for rec in self:
+            user_reviews = rec.review_ids.filtered(
+                lambda r: r.status in ('pending', 'approved') and
+                (r.reviewer_id == self.env.user or
+                 r.reviewer_group_id in self.env.user.groups_id))
+            user_reviews.write({'status': 'rejected'})
+
+    @api.multi
+    def request_validation(self):
+        td_obj = self.env['tier.definition']
+        tr_obj = created_trs = self.env['tier.review']
+        for rec in self:
+            if getattr(rec, self._state_field) in self._state_from:
+                if rec.need_validation:
+                    tier_definitions = td_obj.search([
+                        ('model', '=', self._name)], order="sequence desc")
+                    sequence = 0
+                    for td in tier_definitions:
+                        if self.evaluate_tier(td):
+                            sequence += 1
+                            created_trs += tr_obj.create({
+                                'model': self._name,
+                                'res_id': rec.id,
+                                'definition_id': td.id,
+                                'sequence': sequence,
+                            })
+                    # TODO: notify? post some msg in chatter?
```

Comment:
@naglis FYI we'll need notifications, so question is if we shouldn't add here a func. that we can hook later :-)

Reply:
I don't think that it should be a part of module migration. Please note that  the `12.0-mig-base-tier-validation` branch is the same in the OCA PR.

Reply:
I agree, it's only note for a future work.